specs/deposits.md

```diff
@@ -0,0 +1,272 @@
+```solidity
+    // deposit is stored using arbitrary id
+    mapping(uint256 => Deposit) public deposits;
+```
+
+Deposit ID consists of funder address and nonce merged together into U256 id:
+Nonce is chosen by funder when creating deposit.
+
+```solidity
+    function idFromNonce(uint64 nonce) public view returns (uint256) {
+        return idFromNonceAndFunder(nonce, msg.sender);
+    }
+    function idFromNonceAndFunder(uint64 nonce, address funder) public pure returns (uint256) {
+        return (uint256(uint160(funder)) << 96) ^ uint256(nonce);
+    }
+    function nonceFromId(uint256 id) public pure returns (uint64) {
+        return uint64(id);
+    }
+    function funderFromId(uint256 id) public pure returns (address) {
+        return address(uint160(id >> 96));
+    }
+```
+
+### Additional open questions that need to be answered ASAP
+
+![thinking](deposit_assets/thinking300.webp)
+
+* Should yagna adjust allocation to the data returned in deposit view? 
+* Should yagna allocation be for full amount or only for amount - fee?
+
+Additional notes from Witek:
+https://www.notion.so/golemnetwork/Specifications-f664c647d3d541b1aa2ad0fa98624ed9#20246ad4207b46b0a94d6110a56107c4
+
+### Yagna implementation
+
+Deposit
+```rust
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Deposit {
+    pub id: String,
+    pub contract: String,
+}
+```
+Deposit view (read from contract given using ```getDeposit(uint256 id)```)
+```rust
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct DepositView {
+    pub id: String,
+    pub nonce: u64,
+    pub funder: String,
+    pub spender: String,
+    pub amount: u128,
+    pub fee_amount: u128,
+    pub valid_to: u64,
+}
+```
+
+Api extension of optional field deposit in ya-client
+```yaml
+    Allocation:
+      properties:
+...
+        deposit:
+          type: object
+          properties:
+            id:
+              type: string
+            contract:
+              type: string
+          required:
+            - id
+            - contract
+```
+
+Changes in erc20_payment_lib and yagna has to be made to enable payments directly from deposit
+using methods.
+```
+deposit_single_transfer(id, ...
+deposit_transfer(id, ...
+deposit_single_transfer_and_close(id, ...
+deposit_transfer_and_close(id, ...
+```
+
+### Flow
+
+![flow](deposit_assets/flow300.webp)
+
+1. Funder creates deposit using createDeposit function getting deposit ID.
+2. Funder sends deposit ID to Spender.
+3. Spender uses deposit ID to create allocation on yagna.
+4. Spender start processing tasks (agreements with Providers)
+5. Funder can extend deposit using extendDeposit function.
+6. Spender can amend allocation using deposit ID.
+7. Spender when finished can close deposit using closeDeposit function.
+8. Alternatively if Spender fail to close allocation Funder can terminate deposit using terminateDeposit function after validTo date elapses.
+
+
+### Example usage flow chart
```

Comment:
I would move this up, this would allow reader to better understand general concept